packages.d/_ALWAYS.fixup.add/init.d/20confd.sh

```diff
@@ -11,12 +11,18 @@ try_partition() {
     # Mount partition as read-only
     mount /dev/$1 /mnt -o ro
 
-    # Find conf.d in root of partition and extract in alphabetical order each
-    # tar.gz archive into the ramdisk /etc, overwriting any existing file
+    # Find conf.d in root of partition and process in alphabetical order
     if [ -d "/mnt/$CONFDIR" ]; then
-        for conf in /mnt/$CONFDIR/*.tar.gz; do
-            echo Applying configurations from /dev/$1: $conf
-            tar --warning=no-timestamp --extract -f $conf -C /etc
+        for conf in `ls /mnt/$CONFDIR/*.sh /mnt/$CONFDIR/*.tar.gz`; do
+            if [[ $conf =~ .*.sh ]]; then
+                # Execute each sh script in a new /bin/bash process
+                echo Executing configurations from /dev/$1: $conf
+                /bin/bash $conf
+            elif [[ $conf =~ .*.tar.gz ]]; then
+                # Extract each tar.gz archive into /etc overwriting existing files
+                echo Applying configurations from /dev/$1: $conf
+                tar --warning=no-timestamp --extract -f $conf -C /etc
+	    fi
```

Comment:
Use two for loops - first, copy the config, then run the scripts - this allows the scripts to act on the contents of any config.   This would also allow the removal of the bash magic matching syntax, making this script compatible with real bourne shells again.

Also, use "." or "source" to load the scripts, this allows more flexibility (ie, it would allow changing the init system from systemd) and is faster

Finally, while you are in this file, you should check for errors - if the mount fails, dont do anything else, if running a script fails stop processing that dir, if extracting a tar fails stop processing that dir, if one of the files do not exist dont try to process it.

Are you running tests with an environment where you can see the console error messages?  It is imperative that you have serial console support when making changes at this level.  One way to achieve this is to use the qemu-based test environment.

Reply:
>Use two for loops - first, copy the config, then run the scripts - this allows the scripts to act on the contents of any config.

In this case, would it not be counter-intuitive for `30-bootstrap.sh` to act on contents of `50-node-config-save.tar.gz`? I was trying to preserve the expected behaviour of the ordered processing.

```
     /conf.d/00-tomesh-base.tar.gz
             30-bootstrap.sh
             50-node-config-save.tar.gz
```

Having the explicit rule that scripts run after all archive unpackings sounds reasonable too.

>Also, use "." or "source" to load the scripts

I was hoping to better isolate multiple scripts, so if some env var is set by one it does not affect other ones. Thoughts?

>qemu-based test environment

I am in a virtual env and it won't let me run qemu, but let me try with the chromebook this afternoon.

Reply:
yes, it is slightly more counter intuitive - however I am assuming that in a "normal" environment there will be no scripts - that they are a quick way to add or test hacks - so hopefully most of the time people dont have to worry about the ordering.

While you are right about script isolation, I was deliberately intending this to be able to influence later steps.  by sharing the environment, they can pass calculated values to each other (and on to the rest of the /init.d/*.sh scripts too)

Perhaps you should look at improving your virtual env - or purchasing a serial console cable.  It is not unusual to want to run qemu, but the important thing was more to be able to read the error messages..